src/chains/ethereum/ethereum/src/api.ts

```diff
@@ -2884,6 +2884,206 @@ export default class EthereumApi implements Api {
+      if (rewardPercentiles[0] < 0)
+        throw new Error(`${ERR_INVALID_PERCENTILE} ${rewardPercentiles[0]}`);
+      if (rewardPercentiles[rewardPercentiles.length - 1] > 100)
+        throw new Error(
+          `${ERR_INVALID_PERCENTILE} ${
+            rewardPercentiles[rewardPercentiles.length - 1]
+          }`
+        );
+
+      for (let i = 1; i < rewardPercentiles.length; i++) {
+        if (rewardPercentiles[i] <= rewardPercentiles[i - 1]) {
+          throw new Error(
+            `${ERR_INVALID_PERCENTILE} ${rewardPercentiles[i]} ${
+              rewardPercentiles[i - 1]
+            }`
+          );
+        }
+      }
+
+      reward = new Array(totalBlocks);
+    }
+    // totalBlocks is inclusive of newestBlock
+    const oldestBlockNumber =
+      newestBlockNumber - Quantity.from(totalBlocks - 1).toBigInt();
+
+    let currentBlock: Block;
+    let currentPosition = 0;
+
+    while (currentPosition < totalBlocks) {
+      currentBlock = await blockchain.blocks.get(
+        Quantity.toBuffer(oldestBlockNumber + BigInt(currentPosition))
+      );
+
+      baseFeePerGas[currentPosition] = currentBlock.header.baseFeePerGas;
+
+      const gasUsed = currentBlock.header.gasUsed.toBigInt();
+      const gasLimit = currentBlock.header.gasLimit.toBigInt();
+
+      if (gasUsed === gasLimit) {
+        gasUsedRatio[currentPosition] = 1;
+      } else {
+        gasUsedRatio[currentPosition] = Number(
+          `0.${((gasUsed * PRECISION_BIG_INT) / gasLimit)
+            .toString()
+            .padStart(PAD_PRECISION, "0")}`
+        );
+      }
+
+      // For each percentile, find the cost of the unit of gas at that percentage
+      if (reward) {
+        const transactions = currentBlock.getTransactions();
+
+        // If there are no transactions, all reward percentiles are 0.
+        if (transactions.length === 0) {
+          reward[currentPosition] = rewardPercentiles.map(() => {
+            return Quantity.Zero;
+          });
+        } else {
+          // For all transactions, effectiveGasReward = normalized fee per unit of gas
+          // earned by the miner regardless of transaction type
+          const baseFee = baseFeePerGas[currentPosition].toBigInt();
+
+          const receipts = await Promise.all(
+            transactions.map(async tx => {
+              return blockchain.transactionReceipts.get(tx.hash.toBuffer());
+            })
+          );
+
+          // Effective Reward is the amount paid per unit of gas
+          const effectiveRewardAndGasUsed = transactions
+            .map((tx, idx) => {
+              let effectiveGasReward: bigint;
+              if ("maxPriorityFeePerGas" in tx) {
+                effectiveGasReward = tx.maxFeePerGas.toBigInt() - baseFee;
+
+                const maxPriorityFeePerGas = tx.maxPriorityFeePerGas.toBigInt();
+
+                if (maxPriorityFeePerGas < effectiveGasReward) {
+                  effectiveGasReward = maxPriorityFeePerGas;
+                }
```

Comment:
Here's another opportunity for a BigInt.min() function :)

Reply:
:+1: https://github.com/trufflesuite/ganache/pull/3351/commits/6fbf0c5fbf538c94f8e0455ef1c679425556eb6a